### PR TITLE
Pass MooseApp pointer to NEML2

### LINKED ISSUE #28486: Add MOOSE specific `neml2::get_model` method
### Motivation

We'd like to enable stronger coupling between MOSOE and custom NEML2 models and teh ability of NEML2 models to access certain MOOSE APIs like the `DataFileInterface`. For example ANN based surrogates should be able to access model data stored in application data folders.

### Design

Pass in a pointer to the `MooseApp` as an extra option when building the NEML2 model.

### Impact

Added capability. Sorry, @hugary1995 I need this to enable testing in installed applications and for tighter Bison integration.

### PR DESCRIPTION
**Pull request by dschwen**
Closes #28486

### CODE REVIEW

**Comment by dschwen** [comment]
It looks like NEML2 currently has no method in the options object to check if an option exists.

File: modules/solid_mechanics/src/neml2/utils/NEML2Utils.C [hunk 1]
```diff
@@ -54,6 +55,21 @@ namespace NEML2Utils
...
 #ifdef NEML2_ENABLED
 
+neml2::Model &
+get_model_moose(const std::string & mname,
+                MooseBase * moose_base,
+                bool enable_ad,
+                bool force_create)
+{
+  neml2::OptionSet extra_opts;
+  extra_opts.set<MooseApp *>("_MooseApp") = &moose_base->getMooseApp();
```
Comment (hugary1995) [inline modules/solid_mechanics/src/neml2/utils/NEML2Utils.C:65]:
```suggestion
  extra_opts.set<MooseApp *>("_moose_app") = &moose_base->getMooseApp();
```
to be consistent with neml2 naming conventions.

File: modules/solid_mechanics/test/neml2/src/NEML2TestModel.cxx [hunk 2]
```diff
@@ -37,6 +39,10 @@ NEML2TestModel::NEML2TestModel(const OptionSet & options)
     _sum(declare_output_variable<Scalar>("sum")),
     _product(declare_output_variable<Scalar>("product"))
 {
+  const auto moose_app = options.get<MooseApp *>("_MooseApp");
+  neml_assert(moose_app,
+              "The _MooseApp option was not set. The model should be build using the "
```
Comment (hugary1995) [inline modules/solid_mechanics/test/neml2/src/NEML2TestModel.cxx:44]:
```suggestion
              "The _moose_app option was not set. The model should be built using the "
```

File: modules/solid_mechanics/test/neml2/src/NEML2TestModel.cxx [hunk 3]
```diff
@@ -37,6 +39,10 @@ NEML2TestModel::NEML2TestModel(const OptionSet & options)
     _sum(declare_output_variable<Scalar>("sum")),
     _product(declare_output_variable<Scalar>("product"))
 {
+  const auto moose_app = options.get<MooseApp *>("_MooseApp");
```
Comment (hugary1995) [inline modules/solid_mechanics/test/neml2/src/NEML2TestModel.cxx:42]:
```suggestion
  const auto moose_app = options.get<MooseApp *>("_moose_app");
```

File: modules/solid_mechanics/test/neml2/src/NEML2TestModel.cxx [hunk 4]
```diff
@@ -25,6 +26,7 @@ NEML2TestModel::expected_options()
   // Use AD
   options.set<bool>("_use_AD_first_derivative") = true;
   options.set<bool>("_use_AD_second_derivative") = true;
+  options.set<MooseApp *>("_MooseApp") = nullptr;
```
Comment (hugary1995) [inline modules/solid_mechanics/test/neml2/src/NEML2TestModel.cxx:29]:
```suggestion
  options.set<MooseApp *>("_moose_app") = nullptr;
```

File: modules/solid_mechanics/include/neml2/utils/NEML2Utils.h [hunk 5]
```diff
@@ -43,6 +43,12 @@ namespace NEML2Utils
 
 #ifdef NEML2_ENABLED
 
+/// special version of get_model that adds a MooseApp parameter
+neml2::Model & get_model_moose(const std::string & mname,
+                               MooseBase * moose_base,
+                               bool enable_ad = true,
+                               bool force_create = true);
```
Comment (hugary1995) [inline modules/solid_mechanics/include/neml2/utils/NEML2Utils.h:50]: What's the reason for not directly passing a pointer to MooseApp?

**Review by hugary1995 — CHANGES_REQUESTED** [review]
Looks good to me.

File: modules/solid_mechanics/include/neml2/utils/NEML2Utils.h (hunk 5 above)
Comment (dschwen) [reply modules/solid_mechanics/include/neml2/utils/NEML2Utils.h:50]: Because with this signature the caller can just pas `this` into `get_model_moose`.

**Pull request closed without merging**